### PR TITLE
Language Service host API cleanup

### PR DESCRIPTION
**Pull request by mhegazy**
Few cleanup changes to the host API also adding a helper function to get the default library path for node packages.

### CODE REVIEW

**Comment by mhegazy** [comment]
@csnover this change simplifies the host API, take a look and let me know if you see other inconsistencies. still to come is documentation for these APIs.

File: src/services/services.ts [hunk 1]
```diff
@@ -9,7 +9,7 @@
 /// <reference path='formatting\smartIndenter.ts' />
 
 module ts {
-    export var servicesVersion = "0.4"
+    export var servicesVersion = "0.4.1"
```
Comment (DanielRosenwasser) [inline src/services/services.ts]: Just noticed, put a semicolon.

File: src/services/services.ts [hunk 2]
```diff
@@ -841,26 +841,21 @@ module ts {
...
-    export interface Logger {
-        log(s: string): void;
-        trace(s: string): void;
-        error(s: string): void;
-    }
-
     //
     // Public interface of the host of a language service instance.
     //
-    export interface LanguageServiceHost extends Logger {
```
Comment (DanielRosenwasser) [inline src/services/services.ts]: Ohhh, that's why `Logger` was expected in some places.

File: src/services/services.ts [hunk 3]
```diff
@@ -2081,7 +2044,16 @@ module ts {
             }
 
             // Now create a new compiler
-            program = createProgram(hostfilenames, compilationSettings, createCompilerHost());
+            program = createProgram(hostfilenames, compilationSettings, {
+                getSourceFile: getSourceFile,
```
Comment (DanielRosenwasser) [inline src/services/services.ts]: Use a shorthand.

File: src/services/services.ts [hunk 4]
```diff
@@ -2081,7 +2044,16 @@ module ts {
...
 
             // Now create a new compiler
-            program = createProgram(hostfilenames, compilationSettings, createCompilerHost());
+            program = createProgram(hostfilenames, compilationSettings, {
+                getSourceFile: getSourceFile,
+                getCancellationToken: () => cancellationToken,
+                getCanonicalFileName: filename => useCaseSensitivefilenames ? filename : filename.toLowerCase(),
+                useCaseSensitiveFileNames: () => useCaseSensitivefilenames,
+                getNewLine: () => host.getNewLine ? host.getNewLine() : "\r\n",
+                getDefaultLibFilename: getDefaultLibraryFilename,
```
Comment (DanielRosenwasser) [inline src/services/services.ts]: Shorthand

File: tests/cases/unittests/services/documentRegistry.ts [hunk 5]
```diff
@@ -5,8 +5,8 @@ describe("DocumentRegistry", () => {
         var documentRegistry = ts.createDocumentRegistry();
         var defaultCompilerOptions = ts.getDefaultCompilerOptions();
 
-        var f1 = documentRegistry.acquireDocument("file1.ts", defaultCompilerOptions, ts.ScriptSnapshot.fromString("var x = 1;"), "1", false);
-        var f2 = documentRegistry.acquireDocument("file1.ts", defaultCompilerOptions, ts.ScriptSnapshot.fromString("var x = 1;"), "1", false);
+        var f1 = documentRegistry.acquireDocument("file1.ts", defaultCompilerOptions, ts.ScriptSnapshot.fromString("var x = 1;"), "1");
+        var f2 = documentRegistry.acquireDocument("file1.ts", defaultCompilerOptions, ts.ScriptSnapshot.fromString("var x = 1;"), "1");
```
Comment (DanielRosenwasser) [inline tests/cases/unittests/services/documentRegistry.ts]: Not to be a pain, but could we have `/* version */` for the `"1"` parameters?

File: src/compiler/core.ts [hunk 6]
```diff
@@ -633,6 +633,10 @@ module ts {
         }
     }
 
+    export function getDefaultLibraryFilename(options: CompilerOptions): string {
+        return options.target === ScriptTarget.ES6 ? "lib.es6.d.ts" : "lib.d.ts";
+    }
+
```
Comment (DanielRosenwasser) [inline src/compiler/core.ts]: We use some stuff like this in the harness - @yuit was questioning the logic a week ago. Not sure if the problem persists.

**Comment by DanielRosenwasser** [comment]
Looks good, once the minor changes I've suggested are addressed, :+1:

File: src/services/services.ts (hunk 1 above)
Comment (vladima) [inline src/services/services.ts]: Bumping up the API version will require updating the check on the managed side as well

**Comment by csnover** [comment]
It will be a few days for additional feedback from me. It’s up to you if you want to wait :)

File: src/services/services.ts [hunk 7]
```diff
@@ -5724,6 +5696,23 @@ module ts {
...
+    declare var __dirname: string;
+    declare var module: any;
+
+    /**
+      * Get the path of the default library file (lib.d.ts) as distributed with the typescript
+      * node package.
+      * The functionality is not supported if the ts module is consumed outside of a node module. 
+      */
+    export function getDefaultLibraryFilePath(options: CompilerOptions): string {
+        if (typeof module !== "undefined" && module.exports) {
```
Comment (CyrusNajmabadi) [inline src/services/services.ts]: i'm not sure what this means, or why it is correct.

**Comment by steveluc** [comment]
Good simplification.    Did you check whether WScript will also give you the directory of the executing .js file?

:+1:

File: src/compiler/core.ts (hunk 6 above)
Comment (sheetalkamat) [inline src/compiler/core.ts]: should we be more generic here for future targets? something that converts ScriptTaget into string and use that with lib.scriptTagetString.d.ts ?

File: src/services/services.ts (hunk 7 above)
Comment (sheetalkamat) [inline src/services/services.ts]: +1. not sure what we are checking here and why?

Comment (mhegazy) [inline src/services/services.ts]: checking if we are in node. i am open to suggestions.. this is how we do it in sys anyways.

File: src/compiler/core.ts (hunk 6 above)
Comment (mhegazy) [inline src/compiler/core.ts]: i think we should only have one lib file with some details inside on when declarations are relevant, e.g. decorators/attribute annotations that say if this declaration applies to the current compilation or not.
managing multiple files has been a hassle.

File: src/services/services.ts (hunk 7 above)
Comment (csnover) [inline src/services/services.ts]: If you just want to use `__dirname` then check `typeof __dirname`. This current line isn’t a good check for being in a Node.js environment. You can get the same invalid answer with a `define(function (require, exports, module) {})` AMD wrapper or if someone just pollutes some globals. Probably the “best” check for knowing if you are in an environment that is intentionally trying to quack like Node.js is `process && process.versions && process.versions.node` but in any case actually trying to detect the API you want to use is best.

File: src/compiler/core.ts (hunk 6 above)
Comment (yuit) [inline src/compiler/core.ts]: @mhegazy I agree on that. Also, should we make sure that the target exist first? or we are guarantee to have default target as there was a problem in the harness when target isn't specified

**Comment by JsonFreeman** [comment]
:+1:

File: src/services/services.ts (hunk 3 above)
Comment (mhegazy) [inline src/services/services.ts]: done.

File: src/services/services.ts (hunk 1 above)
Comment (mhegazy) [inline src/services/services.ts]: reverted.

File: src/services/services.ts (hunk 4 above)
Comment (mhegazy) [inline src/services/services.ts]: done.

File: tests/cases/unittests/services/documentRegistry.ts (hunk 5 above)
Comment (mhegazy) [inline tests/cases/unittests/services/documentRegistry.ts]: done.

File: src/services/services.ts (hunk 7 above)
Comment (mhegazy) [inline src/services/services.ts]: changed to check for __dirname

**Pull request merged**